skaffold-plugins-core/src/main/java/com/google/cloud/tools/skaffold/downloader/Downloader.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2018 Google LLC. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.skaffold.downloader;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.io.IOException;
+import java.net.URL;
+import java.net.URLConnection;
+import java.nio.channels.Channels;
+import java.nio.channels.FileChannel;
+import java.nio.channels.ReadableByteChannel;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+
+/** Downloads files. */
+public class Downloader {
+
+  /**
+   * Downloads to a destination file.
+   *
+   * @param url the {@link URL} to download
+   * @param destination the destination file to download to
+   * @return the size of the downloaded contents, or -1 if {@code Content-Length} is unknown and
+   *     thus nothing downloaded
+   * @throws IOException if an I/O exception occurred during the download process
+   */
+  public static long download(URL url, Path destination) throws IOException {
+    return download(url, destination, Long.MAX_VALUE);
```

Comment:
Thinking again, the idiomatic read loop typically uses a reasonable fixed size of a read buffer, rather than attempting to read all at once. I am not sure, but I just wonder if `FileChannel` may internally allocate an byte array of size `MAX_VALUE` upfront in some cases where actual low-level copying is necessary. It seems safe to use a reasonably small chunksize anyway, maybe like 1024k? WDYT?

Reply:
I think it might be good to wrap the `connection.getInputStream()` in a `BufferedInputStream` and delegate the buffer size choice to `BufferedInputStream` instead.

Reply:
Sorry, never mind. I tried to search on the Internet if there would be any problem with passing a very large `count` value to `transferFrom()` but couldn't find any. A lot of people seem to just casually use `Long.MAX_VALUE`. `BufferedInputStream` might not matter.